pkg/elf/elf.go

```diff
@@ -0,0 +1,260 @@
+	}
+	if _, err := elf.readValue(int64(opt.offset), opt.size); err != nil {
+		return err
+	}
+	return nil
+}
+
+// copy the ELF from the reader to the writer, substituting the constants with
+// names specified in 'intOptions' with their corresponding values, and the
+// strings specified in 'strOptions' with their corresponding values.
+//
+// Keys in the 'intOptions' / 'strOptions' maps are case-sensitive.
+func (elf *ELF) copy(w io.Writer, r io.ReadSeeker, intOptions map[string]uint32, strOptions map[string]string) error {
+	if len(intOptions) == 0 && len(strOptions) == 0 {
+		// Copy the remaining portion of the file
+		if _, err := io.Copy(w, r); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	globalOff := uint64(0) // current position in file
+	processedOptions := make(map[string]struct{}, len(intOptions)+len(strOptions))
+processSymbols:
+	for _, symbol := range elf.symbols.sort() {
+		scopedLog := log.WithField("symbol", symbol.name)
+
+		// Figure out the value to substitute
+		var value []byte
+		switch symbol.kind {
+		case symbolUint32:
+			v, exists := intOptions[symbol.name]
+			if exists {
+				value = make([]byte, unsafe.Sizeof(v))
+				elf.ByteOrder.PutUint32(value, v)
+			}
+		case symbolString:
+			v, exists := strOptions[symbol.name]
+			if exists {
+				if uint64(len(v)) != symbol.size {
+					return fmt.Errorf("symbol substitution value %q (len %d) must equal length of symbol name %q (len %d)", v, len(v), symbol.name, len(symbol.name))
+				}
+				value = []byte(v)
+			}
+		}
+		if value == nil {
+			for _, prefix := range ignoredPrefixes {
+				if strings.HasPrefix(symbol.name, prefix) {
+					continue processSymbols
+				}
+			}
+			scopedLog.Warning("Skipping symbol substitution")
+			continue processSymbols
+		}
+
+		// Copy data up until this symbol into the new file;
+		// Write the new value and seek past it.
+		dataToCopy := int64(symbol.offset - globalOff)
+		if _, err := io.CopyN(w, r, dataToCopy); err != nil {
+			return err
+		}
+		if err := binary.Write(w, elf.ByteOrder, value); err != nil {
+			return fmt.Errorf("failed to substitute %s: %s", symbol.name, err)
+		}
+		if _, err := r.Seek(int64(symbol.size), io.SeekCurrent); err != nil {
+			return err
+		}
+		processedOptions[symbol.name] = struct{}{}
+		globalOff = symbol.offset + symbol.size
+	}
+
+	// Check for additional options that weren't applied
+	for symbol := range strOptions {
+		if _, processed := processedOptions[symbol]; !processed {
+			return fmt.Errorf("no such string %q in ELF", symbol)
+		}
+	}
+	for symbol := range intOptions {
+		if _, processed := processedOptions[symbol]; !processed {
+			return fmt.Errorf("no such symbol %q in ELF", symbol)
+		}
+	}
+
+	// Copy the remaining portion of the file
+	if _, err := io.Copy(w, r); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Write the received ELF to a new file at the specified location, with the
+// specified options (indexed by name) substituted in the data section.
```

Comment:
nit: And the specified string options substituted in the strtab.

Reply:
Each map key and value in `strOptions` must be of equal length. I believe I left a warning log somewhere if this is not the case.